### PR TITLE
Client reset

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

Reference to the issue(s) addressed by this PR: #795

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: lib/errors.js [hunk 1]
```diff
@@ -32,3 +32,25 @@ AuthError.__proto__ = Error;
...
+
+// argument: { backupPath: '...', configuration: { ... } }
+function ClientResetError(argument) {
+    const error = Error.call(this, 'ClientResetError');
+
+    this.name = 'ClientResetError';
+    this.message = error.message;
+    this.stack = error.stack;
+
+    this.getBackupPath = function() {
```
Comment (bdash) [inline lib/errors.js]: Why not expose a configuration here like we do for "offline" client reset? That's by the far the easiest way for a client to access the backed-up Realm.

**Comment by blagoev** [comment]
did not see a change in lib/browser. Maybe this will fail RN debug target

**Comment by kneth** [comment]
@blagoev You are 100 % right. I have disabled the test for now, and will return to it later.

**Comment by bmunkholm** [comment]
Fixes #795

File: docs/sync.js [hunk 2]
```diff
@@ -55,7 +55,22 @@ class Sync {
...
-
+    /**
+     * Initiate a client reset. The Realm must be closed prior to the reset.
+     * @param {string} [path] - The Realm to reset.
+     * Throws error if reset is not possible.
+     * @example
+     * {
+     *   const config = { sync: { user, url: 'realm://localhost:9080/~/myrealm' } };
+     *   config.sync.error = (sender, error) => {
+     *     if (error.code === 7) { // 7 -> client reset
```
[the remark this answers is not in the document]
Comment (kneth) [reply docs/sync.js]: We can expand the error with an additional property called `name`.

File: tests/js/session-tests.js [hunk 3]
```diff
@@ -646,4 +646,33 @@ module.exports = {
...
+            return new Promise((resolve, _reject) => {
+                var realm;
+                const config = { sync: { user, url: 'realm://localhost:9080/~/myrealm' } };
+                config.sync.error = (sender, error) => {
+                    try {
+                        TestCase.assertEqual(error.code, 7); // 7 -> client reset
+                        TestCase.assertDefined(error.config);
+                        TestCase.assertNotEqual(error.config.path, '');
+                        const original_path = realm.path;
+                        realm.close();
```
[the remark this answers is not in the document]
Comment (kneth) [reply tests/js/session-tests.js]: I think we can only access `SyncError` and `SyncSession` at that point. That means, it is hard to close the Realm.

If you decide to ignore the client reset (by not providing a callback), the app will continue and you might loose data when the app is restarted.

**Pull request merged**